internal/semaphore/semaphore.go

```diff
@@ -0,0 +1,84 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package semaphore
+
+import (
+	"log"
+	"time"
+)
+
+type Semaphore interface {
+	// Acquire a slot in the semaphore with a timeout. If timeout <=0, the slot would be released immediately
+	Acquire(timeout time.Duration) bool
+	
+	// Release a slot in the semaphore.	
+	Release()
+
+	//GetLimit returns semaphore limit.
+	GetLimit() int
+
+	//GetCount returns current number of occupied slots in semaphore.
+	GetCount() int
+
+	//Done closes down the semaphore
+	Done()
+}
+
+type semaphore struct {
+	slots chan struct{}
+}
+
+func NewSemaphore(limit int) Semaphore {
+	if limit < 0 {
+		log.Panic("E! Semaphore size limit must not be negative")
+	}
+
+	sem := &semaphore{
+		slots: make(chan struct{}, limit),
```

Comment:
I was worried about limit being a big number (<16K) and `slots` taking up an unnecessary amount of memory.
From what I can see though empty struct is 0 bytes, and a channel of them is only 8 bytes no matter what the capacity is.

So I am not worried anymore.